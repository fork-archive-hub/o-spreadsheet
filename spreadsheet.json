{"sheets":[{"name":"Sheet1","cells":{"A21":{"content":"Sheet2 => B2:"},"B2":{"content":"Owl is awesome","style":1},"B4":{"content":"Numbers","style":4},"B21":{"content":"=Sheet2!B2","style":7},"C1":{"content":"CF =42"},"C4":{"content":"12.4"},"C5":{"content":"42"},"C7":{"content":"3"},"B9":{"content":"Formulas","style":5},"C9":{"content":"= SUM ( C4 : C5 )"},"C10":{"content":"=SUM(C4:C7)"},"D10":{"content":"note that C7 is empty"},"C11":{"content":"=-(3 + C7 *SUM(C4:C7))"},"C12":{"content":"=SUM(C9:C11)"},"D12":{"content":"this is a sum of sums"},"B14":{"content":"Errors","style":6},"C14":{"content":"=C14"},"C15":{"content":"=(+"},"C16":{"content":"=C15"},"F2":{"content":"italic blablah","style":2},"F3":{"content":"strikethrough","style":3},"H2":{"content":"merged content"},"C20":{"content":"left","border":1},"E20":{"content":"top","border":2},"G20":{"content":"all","border":3},"K3":{"border":3},"B17":{"content":"=WAIT(1000)"},"G13":{"content":"=A1+A2+A3+A4+A5+A6+A7+A8+A9+A10+A11+A12+A13+A14+A15+A16+A17+A18"},"C23":{"content":"0.43","format":"0.00%"},"C24":{"content":"10","format":"#,##0.00"},"C25":{"content":"10.123","format":"#,##0.00"},"G1":{"content":"CF color scale:"},"G2":{"content":"5"},"G3":{"content":"8"},"G4":{"content":"9"},"G5":{"content":"15"},"G6":{"content":"22"},"G8":{"content":"30"},"B26":{"content":"first dataset"},"C26":{"content":"second dataset"},"B27":{"content":"12"},"B28":{"content":"=floor(RAND()*50)"},"B29":{"content":"=floor(RAND()*50)"},"B30":{"content":"=floor(RAND()*50)"},"B31":{"content":"=floor(RAND()*50)"},"B32":{"content":"=floor(RAND()*50)"},"B33":{"content":"=floor(RAND()*50)"},"B34":{"content":"19"},"B35":{"content":"=floor(RAND()*50)"},"C27":{"content":"=floor(RAND()*50)"},"C28":{"content":"=floor(RAND()*50)"},"C29":{"content":"=floor(RAND()*50)"},"C30":{"content":"=floor(RAND()*50)"},"C31":{"content":"=floor(RAND()*50)"},"C32":{"content":"=floor(RAND()*50)"},"C33":{"content":"=floor(RAND()*50)"},"C34":{"content":"=floor(RAND()*50)"},"C35":{"content":"=floor(RAND()*50)"},"A27":{"content":"Emily Anderson (Emmy)"},"A28":{"content":"Sophie Allen (Saffi)"},"A29":{"content":"Chloe Adams"},"A30":{"content":"Megan Alexander (Meg)"},"A31":{"content":"Lucy Arnold (Lulu)"},"A32":{"content":"Hannah Alvarez"},"A33":{"content":"Jessica Alcock (Jess)"},"A34":{"content":"Charlotte Anaya"},"A35":{"content":"Lauren Anthony"}}}]}