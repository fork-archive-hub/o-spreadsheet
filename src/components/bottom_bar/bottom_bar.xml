<templates>
  <t t-name="o-spreadsheet-BottomBar" owl="1">
    <div
      class="o-spreadsheet-bottom-bar o-two-columns d-flex align-items-center fs-5 border-top overflow-hidden bg-light"
      t-on-click="props.onClick"
      t-ref="bottomBar">
      <div
        class="o-sheet-item o-add-sheet d-flex align-items-center p-2 me-1 cursor-pointer"
        t-att-class="{'disabled': env.model.getters.isReadonly()}"
        t-on-click="addSheet">
        <t t-call="o-spreadsheet-Icon.PLUS"/>
      </div>
      <div
        class="o-sheet-item o-list-sheets d-flex align-items-center p-2 me-1 cursor-pointer"
        t-on-click="listSheets">
        <t t-call="o-spreadsheet-Icon.LIST"/>
      </div>
      <div class="o-all-sheets d-flex align-items-center mw-75 overflow-hidden">
        <t t-foreach="getVisibleSheets()" t-as="sheet" t-key="sheet.id">
          <div
            class="o-sheet-item o-sheet d-flex align-items-center py-2 ps-4 pe-3 border-start cursor-pointer user-select-none text-nowrap"
            t-on-click="(ev) => this.activateSheet(sheet.id, ev)"
            t-on-contextmenu.prevent="(ev) => this.onContextMenu(sheet.id, ev)"
            t-att-title="sheet.name"
            t-att-data-id="sheet.id"
            t-attf-class="{{ (sheet.id === env.model.getters.getActiveSheetId()) ? 'active bg-white shadow text-primary': 'text-secondary' }}">
            <span
              class="o-sheet-name"
              t-esc="sheet.name"
              t-on-dblclick="(ev) => this.onDblClick(sheet.id, ev)"
            />
            <span
              class="o-sheet-icon ms-1"
              t-on-click.stop="(ev) => this.onIconClick(sheet.id, ev)">
              <t t-call="o-spreadsheet-Icon.TRIANGLE_DOWN"/>
            </span>
          </div>
        </t>
      </div>

      <t t-set="selectedStatistic" t-value="getSelectedStatistic()"/>
      <div
        t-if="selectedStatistic !== undefined"
        class="o-selection-statistic d-flex align-items-center h-100 bg-white ms-auto p-3 cursor-pointer user-select-none shadow text-secondary"
        t-on-click="listSelectionStatistics">
        <t t-esc="selectedStatistic"/>
        <span>
          <t t-call="o-spreadsheet-Icon.TRIANGLE_DOWN"/>
        </span>
      </div>

      <Menu
        t-if="menuState.isOpen"
        position="menuState.position"
        menuItems="menuState.menuItems"
        onClose="() => this.menuState.isOpen=false"
      />
    </div>
  </t>
</templates>
