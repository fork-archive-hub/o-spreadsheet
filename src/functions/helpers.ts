// HELPERS

import {
  parseNumber,
  isNumber,
  parseDateTime,
  numberToDate,
  toNativeDate,
  InternalDate,
} from "../helpers/index";
import { _lt } from "../translation";

const expectNumberValueError = (value: string) =>
  _lt(
    `The function [[FUNCTION_NAME]] expects a number value, but '${value}' is a string, and cannot be coerced to a number.`
  );

export function toNumber(value: any): number {
  if (typeNumber(value)) {
    return numberValue(value);
  }
  if (typeof value === "boolean") {
    return value ? 1 : 0;
  }
  if (typeof value === "string") {
    if (isNumber(value) || value === "") {
      return parseNumber(value);
    }
    const date = parseDateTime(value);
    if (date) {
      return date.value;
    }
    throw new Error(expectNumberValueError(value));
  }
  return value || 0;
}

export function strictToNumber(value: any): number {
  if (value === "") {
    throw new Error(expectNumberValueError(value));
  }
  return toNumber(value);
}

export function visitNumbers(args: IArguments | any[], cb: (arg: number) => void): void {
  for (let n of args) {
    if (Array.isArray(n)) {
      for (let i of n) {
        for (let j of i) {
          if (typeNumber(j)) {
            cb(numberValue(j));
          }
        }
      }
    } else {
      cb(strictToNumber(n));
    }
  }
}
export function typeNumber(value: any): boolean {
  // Normal case
  if (typeof value === "number") {
    return true;
  }
  // InternalDate case
  if (typeof value === "object" && value !== null) {
    return true;
  }
  return false;
}

export function numberValue(value: number | InternalDate) {
  return typeof value === "number" ? value : value.value;
}

function visitNumbersTextAs0(args: IArguments | any[], cb: (arg: number) => void): void {
  for (let n of args) {
    if (Array.isArray(n)) {
      for (let i of n) {
        for (let j of i) {
          if (j !== undefined && j !== null) {
            if (typeNumber(j)) {
              cb(numberValue(j));
            } else if (typeof j === "boolean") {
              cb(toNumber(j));
            } else {
              cb(0);
            }
          }
        }
      }
    } else {
      cb(toNumber(n));
    }
  }
}

export function visitAny(arg: any, cb: (a: any) => void): void {
  if (Array.isArray(arg)) {
    for (let col of arg) {
      for (let cell of col) {
        cb(cell);
      }
    }
  } else {
    cb(arg);
  }
}

export function visitAnys(
  args: IArguments | any[],
  rangeCb: (a: any) => boolean,
  argCb: (a: any) => boolean
): void {
  for (let arg of args) {
    if (Array.isArray(arg)) {
      for (let col of arg) {
        for (let cell of col) {
          if (!rangeCb(cell)) return;
        }
      }
    } else {
      if (!argCb(arg)) return;
    }
  }
}

export function reduceArgs<T>(
  args: IArguments | any[],
  cb: (acc: T, a: any) => T,
  initialValue: T
): T {
  let val = initialValue;
  for (let arg of args) {
    visitAny(arg, (a) => {
      val = cb(val, a);
    });
  }
  return val;
}

export function reduceNumbers<T>(
  args: IArguments | any[],
  cb: (acc: T, a: any) => T,
  initialValue: T
): T {
  let val = initialValue;
  visitNumbers(args, (a) => {
    val = cb(val, a);
  });
  return val;
}

export function reduceNumbersTextAs0<T>(
  args: IArguments | any[],
  cb: (acc: T, a: any) => T,
  initialValue: T
): T {
  let val = initialValue;
  visitNumbersTextAs0(args, (a) => {
    val = cb(val, a);
  });
  return val;
}

export function toString(value: any): string {
  if (typeof value === "string") {
    return value;
  }
  if (typeNumber(value)) {
    return numberValue(value).toString();
  }
  if (typeof value === "boolean") {
    return value ? "TRUE" : "FALSE";
  }
  return "";
}

const expectBooleanValueError = (value: string) =>
  _lt(
    `The function [[FUNCTION_NAME]] expects a boolean value, but '${value}' is a text, and cannot be coerced to a number.`
  );

export function toBoolean(value: any): boolean {
  if (typeof value === "boolean") {
    return value;
  }
  if (typeof value === "string") {
    if (value) {
      let uppercaseVal = value.toUpperCase();
      if (uppercaseVal === "TRUE") {
        return true;
      }
      if (uppercaseVal === "FALSE") {
        return false;
      }
      throw new Error(expectBooleanValueError(value));
    } else {
      return false;
    }
  }
  if (typeNumber(value)) {
    return numberValue(value) ? true : false;
  }
  return false;
}

export function strictToBoolean(value: any): boolean {
  if (value === "") {
    throw new Error(expectBooleanValueError(value));
  }
  return toBoolean(value);
}

export function visitBooleans(args: IArguments, cb: (a: boolean) => boolean): void {
  visitAnys(
    args,
    (cell) => {
      if (typeof cell === "boolean") {
        return cb(cell);
      }
      if (typeNumber(cell)) {
        return cb(numberValue(cell) ? true : false);
      }
      return true;
    },
    (arg) => (arg !== null ? cb(strictToBoolean(arg)) : true)
  );
}

export function toDate(date: any): Date {
  if (typeof date === "object" && date !== null) {
    return toNativeDate(date);
  }
  if (typeof date === "string") {
    let result = parseDateTime(date);
    if (result !== null && result.jsDate) {
      return result.jsDate;
    }
  }
  return numberToDate(toNumber(date));
}

// -----------------------------------------------------------------------------
// CRITERION FUNCTIONS
// -----------------------------------------------------------------------------

type Operator = ">" | ">=" | "<" | "<=" | "<>" | "=";
interface Predicate {
  operator: Operator;
  operand: any;
  regexp?: RegExp;
}

function getPredicate(descr: string, isQuery: boolean): Predicate {
  let operator: Operator;
  let operand: any;

  let subString = descr.substring(0, 2);

  if (subString === "<=" || subString === ">=" || subString === "<>") {
    operator = subString;
    operand = descr.substring(2);
  } else {
    subString = descr.substring(0, 1);
    if (subString === "<" || subString === ">" || subString === "=") {
      operator = subString;
      operand = descr.substring(1);
    } else {
      operator = "=";
      operand = descr;
    }
  }

  const date = parseDateTime(operand);
  if (date) {
    operand = date.value;
  } else if (isNumber(operand)) {
    operand = toNumber(operand);
  } else if (operand === "TRUE" || operand === "FALSE") {
    operand = toBoolean(operand);
  }

  const result: Predicate = { operator, operand };

  if (typeof operand === "string") {
    if (isQuery) {
      operand += "*";
    }
    result.regexp = operandToRegExp(operand);
  }

  return result;
}

function operandToRegExp(operand: string): RegExp {
  let exp = "";
  let predecessor = "";
  for (let char of operand) {
    if (char === "?" && predecessor !== "~") {
      exp += ".";
    } else if (char === "*" && predecessor !== "~") {
      exp += ".*";
    } else {
      if (char === "*" || char === "?") {
        //remove "~"
        exp = exp.slice(0, -1);
      }
      if (["^", ".", "[", "]", "$", "(", ")", "*", "+", "?", "|", "{", "}", "\\"].includes(char)) {
        exp += "\\";
      }
      exp += char;
    }
    predecessor = char;
  }
  return new RegExp("^" + exp + "$", "i");
}

function evaluatePredicate(value: any, criterion: Predicate): boolean {
  const { operator, operand } = criterion;
  const val = typeNumber(value) ? numberValue(value) : value;

  if (typeof operand === "number" && operator === "=") {
    return toString(val) === toString(operand);
  }

  if (operator === "<>" || operator === "=") {
    let result: boolean;
    if (typeof val === typeof operand) {
      if (criterion.regexp) {
        result = criterion.regexp.test(val);
      } else {
        result = val === operand;
      }
    } else {
      result = false;
    }
    return operator === "=" ? result : !result;
  }

  if (typeof val === typeof operand) {
    switch (operator) {
      case "<":
        return val < operand;
      case ">":
        return val > operand;
      case "<=":
        return val <= operand;
      case ">=":
        return val >= operand;
    }
  }
  return false;
}

/**
 * Functions used especially for predicate evaluation on ranges.
 *
 * Take ranges with same dimensions and take predicates, one for each range.
 * For (i, j) coordinates, if all elements with coordinates (i, j) of each
 * range correspond to the associated predicate, then the function uses a callback
 * function with the parameters "i" and "j".
 *
 * Syntax:
 * visitMatchingRanges([range1, predicate1, range2, predicate2, ...], cb(i,j), likeSelection)
 *
 * - range1 (range): The range to check against predicate1.
 * - predicate1 (string): The pattern or test to apply to range1.
 * - range2: (range, optional, repeatable) ranges to check.
 * - predicate2 (string, optional, repeatable): Additional pattern or test to apply to range2.
 *
 * - cb(i: number, j: number) => void: the callback function.
 *
 * - isQuery (boolean) indicates if the comparison with a string should be done as a SQL-like query.
 * (Ex1 isQuery = true, predicate = "abc", element = "abcde": predicate match the element),
 * (Ex2 isQuery = false, predicate = "abc", element = "abcde": predicate not match the element).
 * (Ex3 isQuery = true, predicate = "abc", element = "abc": predicate match the element),
 * (Ex4 isQuery = false, predicate = "abc", element = "abc": predicate match the element).
 */
export function visitMatchingRanges(
  args: IArguments | any[],
  cb: (i: number, j: number) => void,
  isQuery: boolean = false
): void {
  const countArg = args.length;

  if (countArg % 2 === 1) {
    throw new Error(
      _lt(`Function [[FUNCTION_NAME]] expects criteria_range and criterion to be in pairs.`)
    );
  }

  const dimRow = args[0].length;
  const dimCol = args[0][0].length;

  let predicates: Predicate[] = [];

  for (let i = 0; i < countArg - 1; i += 2) {
    const criteriaRange = args[i];

    if (criteriaRange.length !== dimRow || criteriaRange[0].length !== dimCol) {
      throw new Error(
        _lt(`Function [[FUNCTION_NAME]] expects criteria_range to have the same dimension`)
      );
    }

    const description = toString(args[i + 1]);
    predicates.push(getPredicate(description, isQuery));
  }

  for (let i = 0; i < dimRow; i++) {
    for (let j = 0; j < dimCol; j++) {
      let validatedPredicates = true;
      for (let k = 0; k < countArg - 1; k += 2) {
        const criteriaValue = args[k][i][j];
        const criterion = predicates[k / 2];
        validatedPredicates = evaluatePredicate(criteriaValue, criterion);
        if (!validatedPredicates) {
          break;
        }
      }
      if (validatedPredicates) {
        cb(i, j);
      }
    }
  }
}

// -----------------------------------------------------------------------------
// COMMON FUNCTIONS
// -----------------------------------------------------------------------------

/**
 * Perform a dichotomic search and return the index of the nearest match less than
 * or equal to the target. If all values in the range are greater than the target,
 * -1 is returned.
 * If the range is not in sorted order, an incorrect value might be returned.
 *
 * Example:
 * - [3, 6, 10], 3 => 0
 * - [3, 6, 10], 6 => 1
 * - [3, 6, 10], 9 => 1
 * - [3, 6, 10], 42 => 2
 * - [3, 6, 10], 2 => -1
 */
export function dichotomicPredecessorSearch(range: any[], target: any): number {
  const typeofTarget = typeNumber(target) ? "number" : typeof target;
  const targetValue = typeNumber(target) ? numberValue(target) : target;
  let min = 0;
  let max = range.length - 1;
  let avg = Math.ceil((min + max) / 2);
  let current = range[avg];
  current = typeNumber(current) ? numberValue(current) : current;
  while (max - min > 0) {
    if (typeofTarget === typeof current && current <= targetValue) {
      min = avg;
    } else {
      max = avg - 1;
    }
    avg = Math.ceil((min + max) / 2);
    current = range[avg];
    current = typeNumber(current) ? numberValue(current) : current;
  }
  if (targetValue < current || typeofTarget !== typeof current) {
    // all values in the range are greater than the target, -1 is returned.
    return -1;
  }
  return avg;
}

/**
 * Perform a dichotomic search and return the index of the nearest match more than
 * or equal to the target. If all values in the range are smaller than the target,
 * -1 is returned.
 * If the range is not in sorted order, an incorrect value might be returned.
 *
 * Example:
 * - [10, 6, 3], 3 => 2
 * - [10, 6, 3], 6 => 1
 * - [10, 6, 3], 9 => 0
 * - [10, 6, 3], 42 => -1
 * - [10, 6, 3], 2 => 2
 */
export function dichotomicSuccessorSearch(range: any[], target: any): number {
  const typeofTarget = typeNumber(target) ? "number" : typeof target;
  const targetValue = typeNumber(target) ? numberValue(target) : target;
  let min = 0;
  let max = range.length - 1;
  let avg = Math.floor((min + max) / 2);
  let current = range[avg];
  current = typeNumber(current) ? numberValue(current) : current;
  while (max - min > 0) {
    if (typeofTarget === typeof current && targetValue >= current) {
      max = avg;
    } else {
      min = avg + 1;
    }
    avg = Math.floor((min + max) / 2);
    current = range[avg];
    current = typeNumber(current) ? numberValue(current) : current;
  }
  if (targetValue > current || typeofTarget !== typeof current) {
    return avg - 1;
  }
  return avg;
}
