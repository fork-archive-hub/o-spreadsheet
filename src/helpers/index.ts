export * from "./coordinates";
export * from "./dates";
export * from "./misc";
export * from "./zones";
export * from "./color";
export * from "./numbers";
export * from "./grid_manipulation";
export * from "./uuid";
