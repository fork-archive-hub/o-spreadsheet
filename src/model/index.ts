// history: managed by Core/BasePlugin

// CorePlugins

//             grid??cols/row manipulation, resizing, add/delete row/col
//                 delete/remove col row
//                 reorder row colresize
// selectionclipboard
// multiple sheetsmerge
//             style
// evaluationborder
// historycharts
// sheetentity	addCell
//     width	removeCell
//     height	cellState
//     cells	import
//     import { selectionZoneXC } from "./src/model/selection"
//                         export
//                         *selection
//                         undo/redo
//                         activateSheet
//                         createSheet
//                         renameSheet

/**
 * Plugins
 *
 * core: evaluation, add cell, remove cell, multiple sheet
 *
 *
 */

// base_plugin.ts
// grid_model.ts
// plugin_core.ts
// plugin_grid.ts
// plugin_clipboard
// plugin_merge
// plugin_style
// plugin_border
// plugin_charts

export { GridModel } from "./grid_model";
