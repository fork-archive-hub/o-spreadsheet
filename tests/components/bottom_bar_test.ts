import * as owl from "@odoo/owl";
import { BottomBar } from "../../src/components/bottom_bar";
import { Menus, OpenMenuEvent } from "../../src/components/spreadsheet";
import { Model } from "../../src/model";
import { CommandResult } from "../../src/types";
import { triggerMouseEvent } from "../dom_helper";
import { makeTestFixture, mockUuidV4To, nextTick, SpreadSheetParent } from "../helpers";
import { simulateContextMenu } from "./context_menu_test";
const { Component, useState, hooks, tags } = owl;
const { useExternalListener } = owl.hooks;
jest.mock("../../src/helpers/uuid", () => require("../__mocks__/uuid"));

const { xml } = tags;
const { useSubEnv } = hooks;

let fixture: HTMLElement;
let model: Model;
let parent: Parent | SpreadSheetParent;
class Parent extends Component<any, any> {
  static template = xml`
  <div class="parent" t-on-open-menu="openMenu">
    <BottomBar menuIsOpen="menu.isOpen==='bottomBarMenu'"/>
  </div>`;

  static components = { BottomBar };
  model: Model;
  menu = useState({ isOpen: "" } as {
    isOpen: "" | Menus;
  });
  constructor(model: Model) {
    super();
    useSubEnv({
      openSidePanel: (panel: string) => {},
      dispatch: jest.fn(() => ({ status: "SUCCESS" } as CommandResult)),
      getters: model.getters,
      _t: (s: string) => s,
      askConfirmation: jest.fn(),
    });
    useExternalListener(window as any, "click", this.onClick);
    this.model = model;
  }
  mounted() {
    this.model.on("update", this, this.render);
  }

  willUnmount() {
    this.model.off("update", this);
  }

  openMenu(ev: OpenMenuEvent) {
    this.menu.isOpen = ev.detail.menu;
  }

  onClick() {
    this.menu.isOpen = "";
  }
}

describe("Bottombar", () => {
  beforeEach(async () => {
    fixture = makeTestFixture();
    model = new Model();
    parent = new Parent(model);
    await parent.mount(fixture);
  });

  afterEach(() => {
    fixture.remove();
    parent.destroy();
  });

  test("simple rendering", async () => {
    expect(fixture.querySelector(".o-spreadsheet-bottom-bar")).toMatchSnapshot();
  });

  test("Can create a new sheet", async () => {
    mockUuidV4To(42);
    triggerMouseEvent(".o-add-sheet", "click");
    expect(parent.env.dispatch).toHaveBeenCalledWith("CREATE_SHEET", {
      activate: true,
      id: "42",
    });
  });

  test("Can activate a sheet", async () => {
    triggerMouseEvent(".o-sheet", "click");
    const from = parent.model.getters.getActiveSheet();
    const to = from;
    expect(parent.env.dispatch).toHaveBeenCalledWith("ACTIVATE_SHEET", { from, to });
  });

  test("Can rename a sheet with dblclick", async () => {
    triggerMouseEvent(".o-sheet-name", "dblclick");
    await nextTick();
    const sheet = model.getters.getActiveSheet();
    expect(parent.env.dispatch).toHaveBeenCalledWith("RENAME_SHEET", {
      sheet,
      interactive: true,
    });
  });

  test("Can open context menu of a sheet", async () => {
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(0);
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(1);
  });

  test("Can open context menu of a sheet with the arrow", async () => {
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(0);
    triggerMouseEvent(".o-sheet-icon", "click");
    await nextTick();
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(1);
  });

  test("Click on the arrow when the context menu is open should close it", async () => {
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(0);
    triggerMouseEvent(".o-sheet-icon", "click");
    await nextTick();
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(1);
    triggerMouseEvent(".o-sheet-icon", "click");
    await nextTick();
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(0);
  });

  test("Can move right a sheet", async () => {
    model.dispatch("CREATE_SHEET", { id: "42" });
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    const sheet = model.getters.getActiveSheet();
    triggerMouseEvent(".o-menu-item[data-name='move_right'", "click");
    expect(parent.env.dispatch).toHaveBeenCalledWith("MOVE_SHEET", { sheet, direction: "right" });
  });

  test("Can move left a sheet", async () => {
    model.dispatch("CREATE_SHEET", { id: "42", activate: true });
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    const sheet = model.getters.getActiveSheet();
    triggerMouseEvent(".o-menu-item[data-name='move_left'", "click");
    expect(parent.env.dispatch).toHaveBeenCalledWith("MOVE_SHEET", { sheet, direction: "left" });
  });

  test("Move right and left are not visible when it's not possible to move", async () => {
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    expect(fixture.querySelector(".o-menu-item[data-name='move_left'")).toBeNull();
    expect(fixture.querySelector(".o-menu-item[data-name='move_right'")).toBeNull();
  });

  test("Can rename a sheet", async () => {
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    const sheet = model.getters.getActiveSheet();
    triggerMouseEvent(".o-menu-item[data-name='rename'", "click");
    expect(parent.env.dispatch).toHaveBeenCalledWith("RENAME_SHEET", {
      sheet,
      interactive: true,
    });
  });

  test("Can duplicate a sheet", async () => {
    mockUuidV4To(123);
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    const sheet = model.getters.getActiveSheet();
    const name = `Copy of ${model.getters.getSheets()[0].name}`;
    triggerMouseEvent(".o-menu-item[data-name='duplicate'", "click");
    expect(parent.env.dispatch).toHaveBeenCalledWith("DUPLICATE_SHEET", {
      sheet,
      id: "123",
      name,
    });
  });

  test("Can delete a sheet", async () => {
    model.dispatch("CREATE_SHEET", { id: "42" });
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    const sheet = model.getters.getActiveSheet();
    triggerMouseEvent(".o-menu-item[data-name='delete'", "click");
    expect(parent.env.dispatch).toHaveBeenCalledWith("DELETE_SHEET_CONFIRMATION", { sheet });
  });

  test("Delete sheet is not visible when there is only one sheet", async () => {
    triggerMouseEvent(".o-sheet", "contextmenu");
    await nextTick();
    expect(fixture.querySelector(".o-menu-item[data-name='delete'")).toBeNull();
  });

  test("Can open the list of sheets", async () => {
    const sheet = model.getters.getActiveSheet();
    model.dispatch("CREATE_SHEET", { id: "42" });
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(0);
    triggerMouseEvent(".o-list-sheets", "click");
    await nextTick();
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(1);
    const sheets = fixture.querySelectorAll(".o-menu-item");
    expect(sheets.length).toBe(2);
    expect((sheets[0] as HTMLElement).dataset.name).toBe(sheet);
    expect((sheets[1] as HTMLElement).dataset.name).toBe("42");
  });

  test("Can activate a sheet from the list of sheets", async () => {
    const sheet = model.getters.getActiveSheet();
    model.dispatch("CREATE_SHEET", { id: "42" });
    triggerMouseEvent(".o-list-sheets", "click");
    await nextTick();
    triggerMouseEvent(".o-menu-item[data-name='42'", "click");
    expect(parent.env.dispatch).toHaveBeenCalledWith("ACTIVATE_SHEET", { from: sheet, to: "42" });
  });
});

describe("Spreadsheet menu interactions", () => {
  beforeEach(async () => {
    fixture = makeTestFixture();
    model = new Model();
    parent = new SpreadSheetParent(model);
    await parent.mount(fixture);
  });

  afterEach(() => {
    fixture.remove();
    parent.destroy();
  });
  test("opening topbar tool menu closes bottom bar menu", async () => {
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(0);
    expect(fixture.querySelectorAll(".o-dropdown-content").length).toBe(0);
    triggerMouseEvent(".o-sheet", "contextmenu"); //opens bottombarMenu
    await nextTick();
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(1);
    expect(fixture.querySelectorAll(".o-dropdown-content").length).toBe(0);
    fixture.querySelector('span[title="Borders"]')!.dispatchEvent(new Event("click")); //opens topbar tool menu
    await nextTick();
    expect(fixture.querySelectorAll(".o-dropdown-content").length).toBe(1);
    expect(fixture.querySelectorAll(".o-menu")).toHaveLength(0);
  });

  test("opening topbar context menu closes bottom bar menu", async () => {
    expect(fixture.querySelectorAll(".o-spreadsheet-bottom-bar .o-menu")).toHaveLength(0);
    expect(fixture.querySelectorAll(".o-topbar-topleft .o-menu")).toHaveLength(0);
    triggerMouseEvent(".o-sheet", "contextmenu"); //opens bottombarMenu
    await nextTick();
    expect(fixture.querySelectorAll(".o-spreadsheet-bottom-bar .o-menu")).toHaveLength(1);
    expect(fixture.querySelectorAll(".o-topbar-topleft .o-menu")).toHaveLength(0);
    triggerMouseEvent(".o-topbar-menu[data-id='file']", "click"); //opens topbar context menu
    await nextTick();
    expect(fixture.querySelectorAll(".o-spreadsheet-bottom-bar .o-menu")).toHaveLength(0);
    expect(fixture.querySelectorAll(".o-topbar-topleft .o-menu")).toHaveLength(1);
  });

  test("opening grid context menu closes bottom bar menu", async () => {
    expect(fixture.querySelectorAll(".o-spreadsheet-bottom-bar .o-menu")).toHaveLength(0);
    expect(fixture.querySelectorAll(".o-grid-context-menu .o-menu")).toHaveLength(0);
    triggerMouseEvent(".o-sheet", "contextmenu"); //opens bottombarMenu
    await nextTick();
    expect(fixture.querySelectorAll(".o-spreadsheet-bottom-bar .o-menu")).toHaveLength(1);
    expect(fixture.querySelectorAll(".o-grid-context-menu .o-menu")).toHaveLength(0);
    simulateContextMenu(300, 200); //grid context menu
    await nextTick();
    expect(fixture.querySelectorAll(".o-spreadsheet-bottom-bar .o-menu")).toHaveLength(0);
    expect(fixture.querySelectorAll(".o-grid-context-menu .o-menu")).toHaveLength(1);
  });
});
