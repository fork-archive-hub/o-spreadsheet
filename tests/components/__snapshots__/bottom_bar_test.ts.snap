// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bottombar simple rendering 1`] = `
<div
  class="o-spreadsheet-bottom-bar"
>
  <div
    class="o-sheet-item o-add-sheet"
  >
    <svg
      class="o-icon"
    >
      <path
        d="M8,0 L10,0 L10,8 L18,8 L18,10 L10,10 L10,18 L8,18 L8,10 L0,10 L0,8 L8,8"
        fill="#000000"
      />
    </svg>
  </div>
  <div
    class="o-sheet-item o-list-sheets"
  >
    <svg
      class="o-icon"
      viewBox="0 0 384 384"
    >
      <rect
        height="42.667"
        width="384"
        x="0"
        y="277.333"
      />
      <rect
        height="42.667"
        width="384"
        x="0"
        y="170.667"
      />
      <rect
        height="42.667"
        width="384"
        x="0"
        y="64"
      />
    </svg>
  </div>
  <div
    class="o-all-sheets"
  >
    <div
      class="o-sheet-item o-sheet active"
      data-id="89"
      title="Sheet1"
    >
      <span
        class="o-sheet-name"
      >
        Sheet1
      </span>
      <span
        class="o-sheet-icon"
      >
        <svg
          class="o-icon"
        >
          <polygon
            fill="#000000"
            points="0 0 4 4 8 0"
            transform="translate(5 7)"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`Context Menu context menu simple rendering 1`] = `
<div
  class="o-menu"
  style="top:200px;left:300px;max-height:1000px"
>
  <div
    class="o-menu-item"
    data-name="cut"
    title="Cut"
  >
    Cut
  </div>
  <div
    class="o-menu-item"
    data-name="copy"
    title="Copy"
  >
    Copy
  </div>
  <div
    class="o-menu-item"
    data-name="paste"
    title="Paste"
  >
    Paste
  </div>
  <div
    class="o-menu-item o-menu-root o-separator"
    data-name="paste_special"
    title="Paste special"
  >
    Paste special
    <svg
      class="o-icon"
    >
      <polygon
        fill="#000000"
        points="0 0 4 4 0 8"
        transform="translate(5 3)"
      />
    </svg>
  </div>
  <div
    class="o-menu-item"
    data-name="add_row_before"
    title="Insert row"
  >
    Insert row
  </div>
  <div
    class="o-menu-item o-separator"
    data-name="add_column_before"
    title="Insert column"
  >
    Insert column
  </div>
  <div
    class="o-menu-item"
    data-name="delete_row"
    title="Delete row 8"
  >
    Delete row 8
  </div>
  <div
    class="o-menu-item o-separator"
    data-name="delete_column"
    title="Delete column C"
  >
    Delete column C
  </div>
  <div
    class="o-menu-item disabled"
    data-name="clear_cell"
    title="Clear cell"
  >
    Clear cell
  </div>
  <div
    class="o-menu-item"
    data-name="conditional_formatting"
    title="Conditional formatting"
  >
    Conditional formatting
  </div>
</div>
`;
